### PR TITLE
Drop numpy requirement 

### PR DESCRIPTION
**Pull request by Magi3r**
I refactored the code to do what the numpy functions did with builtin functions.
Numpy accepted anything that could somehow be interpreted as an array, now the `set_image` Function really needs an python-list or int, although the list can be nested in any way.

The given examples all work with this changes, after refactoring them to call set_image with python lists instead of numpy-arrays.

### CODE REVIEW

File: pyghthouse/data/canvas.py [hunk 1]
```diff
@@ -1,26 +1,90 @@
 from typing import Union, Iterable
-import numpy as np
-
-
+# import numpy as np
```
Comment (fwcd) [inline pyghthouse/data/canvas.py]: I think we can remove this:

```suggestion
```

File: pyghthouse/data/canvas.py [hunk 2]
```diff
@@ -1,26 +1,90 @@
...
+            self.image[:][:][0] = 255
         else:
             self.set_image(initial_image)
 
-    def set_image(self, new_image: VALID_IMAGE_TYPE) -> np.array:
+    """
+    Flattens the list. Works for nested lists too.
+    """
+    def flatten_list(self,element):
+        if(type(element)==list):
```
Comment (fwcd) [inline pyghthouse/data/canvas.py]:
```suggestion
        if isinstance(element, list):
```

File: pyghthouse/data/canvas.py [hunk 3]
```diff
@@ -1,26 +1,90 @@
...
+
+    """
+    Parses user input to image data.
+    new_image: int | list
+        If int provided, sets all values to this number. Must be between 0 and 255.
+        If list provided, it must contain exactly as many elements as the image needs values,
+            but it may be nested in any way.
+    """
+    def new_image_to_image_array(self,new_image: VALID_IMAGE_TYPE) -> list:
+        if(type(new_image)==int):
```
Comment (fwcd) [inline pyghthouse/data/canvas.py]:
```suggestion
        if isinstance(new_image, int):
```

File: pyghthouse/data/canvas.py [hunk 4]
```diff
@@ -1,26 +1,90 @@
...
+            if(0<=new_image<=255):
```
Comment (fwcd) [inline pyghthouse/data/canvas.py]: Interesting, I didn't know Python supported this kind of multi-way comparison. :D

Style nit: Please omit parentheses whenever possible

```suggestion
            if 0 <= new_image <= 255:
```

File: pyghthouse/data/canvas.py [hunk 5]
```diff
@@ -1,26 +1,90 @@
...
+                return self.init_image_array(lambda: new_image)
+            else:
+                raise ValueError("Color must be 0<=color<=255")
+            
+        if(type(new_image)==list):
```
Comment (fwcd) [inline pyghthouse/data/canvas.py]:
```suggestion
        if isinstance(new_image, list):
```

File: pyghthouse/data/canvas.py [hunk 6]
```diff
@@ -1,26 +1,90 @@
...
+        return list
+
+    """
+    Parses user input to image data.
+    new_image: int | list
+        If int provided, sets all values to this number. Must be between 0 and 255.
+        If list provided, it must contain exactly as many elements as the image needs values,
+            but it may be nested in any way.
+    """
+    def new_image_to_image_array(self,new_image: VALID_IMAGE_TYPE) -> list:
```
Comment (fwcd) [inline pyghthouse/data/canvas.py]: Please make sure to specify the type arguments:

```suggestion
    def new_image_to_image_array(self,new_image: VALID_IMAGE_TYPE) -> list[int]:
```

File: pyghthouse/data/canvas.py [hunk 7]
```diff
@@ -1,26 +1,90 @@
...
-    VALID_IMAGE_TYPE = Union[np.ndarray, Iterable, int]
+    VALID_IMAGE_TYPE = Union[Iterable, int]
     IMAGE_SHAPE = (14, 28, 3)
 
     def __init__(self, initial_image=None):
-        self.image = np.zeros(self.IMAGE_SHAPE, dtype=np.ubyte)
+        self.image = self.init_image_array()
         if initial_image is None:
-            self.image[:, :, 0] = 255
+            self.image[:][:][0] = 255
```
Comment (fwcd) [inline pyghthouse/data/canvas.py:11]: One design question would be if we want to use nested lists. While it's certainly convenient (especially considering that the library is targeted as beginners), we're also sending images rather frequently, so the overhead of using doubly-nested arrays might become noticeable. 1D lists would likely be more efficient, but would become a bit tedious to use.

Numpy arrays are actually kind of nice in that regard that they expose a high-level API (n-dimensional indexing), while using rather efficient storage.

The more I think about it, the more I'm wondering if there really is a big downside to having numpy as a dependency. It's a pretty standard library across the ecosystem, not that heavyweight and the dependency issues should be solved by using looser version bounds (see #2).

An alternative approach would perhaps be that we _accept_ both lists and numpy arrays  (e.g. by using `np.asarray`), while still using numpy arrays internally. That would make the API easy-to-use for everyone, but also let advanced users continue to pass numpy arrays for efficiency. What do you think?

**Review by fwcd — COMMENTED** [review]
Thanks for the PR, I've left some thoughts below!

File: pyghthouse/data/canvas.py (hunk 6 above)
Comment (Magi3r) [reply pyghthouse/data/canvas.py]: This implementation supports nested lists and all datatypes that can be cast to integers. Thats why im not that sure about list[int]. On the other side, this is python and the type hints are ignored by the interpreter anyway.

Comment (fwcd) [reply pyghthouse/data/canvas.py]: Then I think it should be `list[Any]`. If you use VSCode I recommend trying to enable strict type checking in the settings, which should highlight these kinds of issues. But to be fair, I don't know how much of the remaining codebase is already passing the strict check.

File: examples/rgbscan.py [hunk 8]
```diff
@@ -6,13 +6,13 @@
 
 def image_gen():
     image = np.zeros((14, 28, 3))
-    yield image
+    yield image.tolist()
     while True:
         for j in range(3):
             for x in range(14):
                 for y in range(28):
```
Comment (fwcd) [inline examples/rgbscan.py:13]: Btw perhaps we could refactor this to use constants? (but that's out of scope for the PR)

File: pyghthouse/data/canvas.py (hunk 1 above)
Comment (fwcd) [reply pyghthouse/data/canvas.py]: Btw, feel free to close any discussions that are implemented!

File: examples/twopoints.py [hunk 9]
```diff
@@ -55,7 +55,7 @@ def callback(self):
                        fill=tuple(from_hsv(self.hue, 1, 1)))
         output = self.img.copy()
         output.thumbnail((28, 14))
-        return np.asarray(output)
+        return np.asarray(output).tolist()
```
Comment (fwcd) [inline examples/twopoints.py:58]: I'm not sure if we can do this if we don't require numpy. Then again, this script requires Pillow, so we're relying on implicit dependencies one way or the other. Off top of my head I think we should be able to do

```suggestion
        return output.tolist()
```

**Pull request closed without merging**